mono/profiler/proflog.c

```diff
@@ -2007,101 +2015,104 @@ typedef struct {
 } AsyncFrameInfo;
 
 typedef struct {
+	MonoLockFreeQueueNode node;
+	MonoProfiler *prof;
+	uint64_t elapsed;
+	uintptr_t tid;
+	void *ip;
 	int count;
-	AsyncFrameInfo *data;
-} AsyncFrameData;
+	AsyncFrameInfo frames [MONO_ZERO_LEN_ARRAY];
+} SampleHit;
 
 static mono_bool
 async_walk_stack (MonoMethod *method, MonoDomain *domain, void *base_address, int offset, void *data)
 {
-	AsyncFrameData *frame = (AsyncFrameData *)data;
-	if (frame->count < num_frames) {
-		frame->data [frame->count].method = method;
-		frame->data [frame->count].domain = domain;
-		frame->data [frame->count].base_address = base_address;
-		frame->data [frame->count].offset = offset;
-		// printf ("In %d at %p (dom %p) (native: %p)\n", frame->count, method, domain, base_address);
-		frame->count++;
+	SampleHit *sample = (SampleHit *) data;
+
+	if (sample->count < num_frames) {
+		int i = sample->count;
+
+		sample->frames [i].method = method;
+		sample->frames [i].domain = domain;
+		sample->frames [i].base_address = base_address;
+		sample->frames [i].offset = offset;
+
+		sample->count++;
 	}
-	return frame->count == num_frames;
+
+	return sample->count == num_frames;
 }
 
-/*
-(type | frame count), tid, time, ip, [method, domain, base address, offset] * frames
-*/
-#define SAMPLE_EVENT_SIZE_IN_SLOTS(FRAMES) (4 + (FRAMES) * 4)
+#define SAMPLE_SLOT_SIZE(FRAMES) (sizeof (SampleHit) + sizeof (AsyncFrameInfo) * (FRAMES - MONO_ZERO_LEN_ARRAY))
+#define SAMPLE_BLOCK_SIZE (mono_pagesize ())
+
+static void
+reuse_sample_hit (gpointer p)
+{
+	SampleHit *sample = p;
+
+	mono_lock_free_queue_node_unpoison (&sample->node);
+	mono_lock_free_queue_enqueue (&sample->prof->sample_reuse_queue, &sample->node);
+}
 
 static void
 mono_sample_hit (MonoProfiler *profiler, unsigned char *ip, void *context)
 {
-	StatBuffer *sbuf;
-	AsyncFrameInfo frames [num_frames];
-	AsyncFrameData bt_data = { 0, &frames [0]};
-	uint64_t now;
-	uintptr_t *data, *new_data, *old_data;
-	uintptr_t elapsed;
-	int timedout = 0;
-	int i;
 	if (in_shutdown)
 		return;
-	now = current_time ();
 
-	mono_stack_walk_async_safe (&async_walk_stack, context, &bt_data);
+	InterlockedIncrement (&sample_hits);
+
+	uint64_t now = current_time ();
+	SampleHit *sample;
+
+again:
+	sample = (SampleHit *) mono_lock_free_queue_dequeue (&profiler->sample_reuse_queue);
+	gboolean hazardous;
+
+	if (sample && (hazardous = mono_thread_is_pointer_hazardous (sample))) {
+		mono_lock_free_queue_node_init (&sample->node, TRUE);
+		mono_thread_hazardous_try_free (sample, reuse_sample_hit);
+
+		InterlockedIncrement (&sample_hazards);
+
+		// Hazards are very rare, so just try again.
+		goto again;
+	}
+
+	if (!sample || hazardous) {
+		if (InterlockedRead (&sample_allocations) >= max_allocated_sample_hits)
+			return;
```

Comment:
I do think in addition to just returning, we should increment a "dropped frames due to insufficient reusable sample events" counter at this place. Being able to quantify frame drop would be helpful for a user trying to pick a correct value for maxsamples=. Also, I think it would be helpful to have a comment on this block explaining what is happening ("No SampleHits are available for reuse. Allocate a new one if we are allowed and drop the frame otherwise.", for example), the fact this `return` is dropping the frame on the floor is very important but may not be obvious on a skim of the code.

Also: Why are you checking `|| hazardous` in the condition if the previous clause prevents flow control from reaching this point when `hazardous` is true?